cpp/src/utilities/type_dispatcher.hpp

```diff
@@ -16,20 +16,186 @@
 #ifndef TYPE_DISPATCHER_HPP
 #define TYPE_DISPATCHER_HPP
 
-#include "wrapper_types.hpp"
-#include "release_assert.cuh"
+#include <utilities/wrapper_types.hpp>
+#include <utilities/release_assert.cuh>
 
 #include <cudf/types.h>
 
 // Forward decl
 class NVStrings;
 
-#include <cassert>
 #include <utility>
+#include <type_traits>
+
+namespace cudf {
+
+namespace detail {
+
+template <typename... List> struct list { };
```

Comment:
Would love a summary documentation block on how all of this works.

Reply:
@jrhemstad : The documentation might be more convoluted and less readable than the code...

Reply:
If it can't be easily described, how is anyone else supposed to understand it? 

Some libcudf developers aren't experienced with advanced TMP, and it would be helpful to at least provide an outline/summary of the strategy here. It does not need to be rigorously detailed.

Reply:
I said it can't be more easily described than by the code itself. But I will write something.